File: Week03/python/demo_class.py
# Uncomment each block to test the individual cases.

# Block 1
# data = [12]
# backup = data
# data[0] = 11
# print(backup[0])
# print(data is backup)

# Block 2
# data = [12]
# backup = [0]
# data[0] = 11
# print(data == backup)

# Block 3
# data = [1, 2]
# backup = [12]
# print(data == backup)

# Block 4
# data = [[1, 2], [3, 4]]
# backup = [[1, 2], [3, 4]]
# print(data is backup)
# print(data == backup)
# print(all(x == y for x, y in zip(data, backup)))
# print(data == backup)

# Block 5
# data = [1, 2]
# backup = data
# data[0] = 11
# print(backup[0])

# print(data is backup)
# print(data == backup)
# print(data == backup)

# Block 6
# data = [1, 2]
# backup = data.copy()
# data[0] = 11
# print(backup[0])

# print(data is backup)
# print(data == backup)
# print(data == backup)

# Block 7
# data = [[1, 2], [3, 4]]
# backup = copy.deepcopy(data)
# data[0][0] = 11
# print(backup[0][0])

# print(data is backup)
# print(data == backup)
# print(all(x == y for x, y in zip(data, backup)))
# print(data == backup)

# Block 8
data = [[1, 2], [3, 4]]
backup = [row.copy() for row in data]
data[0][0] = 11
print(backup[0][0])

print(data is backup)
print(data == backup)
print(all(x == y for x, y in zip(data, backup)))
# print(data == backup) # This line is similar to Arrays.deepEquals in Java

File: linked_list/java/DemoClass.java
import java.util.Arrays;

public class DemoClass {
    public static void main(String[] args) {
        // int[] data = { 12 };
        // int[] backup;
        // backup = data;
        // data[0] = 11;
        // System.out.println(backup[0]);
        // System.out.println(data == backup);

        // int[] data = { 12 };
        // int[] backup = new int[1];
        // data[0] = 11;
        // System.out.println(data.equals(backup));

        // int[] data = { 1,2 };
        // int[] backup = new int[1];
        // backup[0] = 12;
        // System.out.println(Arrays.equals(data, backup));

        // int[][] data = { { 1, 2 }, { 3, 4 } };
        // int[][] backup = new int[2][2];
        // backup[0][0] = 1;
        // backup[0][1] = 2;
        // backup[1][0] = 3;
        // backup[1][1] = 4;
        // System.out.println(data == backup);
        // System.out.println(data.equals(backup));
        // System.out.println(Arrays.equals(data, backup));
        // System.out.println(Arrays.deepEquals(data, backup));

        // int[] data = { 1, 2 };
        // int[] backup;
        // backup = data;
        // data[0] = 11;
        // System.out.println(backup[0]);

        // System.out.println(data == backup);
        // System.out.println(data.equals(backup));
        // System.out.println(Arrays.equals(data, backup));
        // // System.out.println(Arrays.deepEquals(data, backup));

        // int[] data = { 1, 2 };
        // int[] backup;
        // backup = data.clone();
        // data[0] = 11;
        // System.out.println(backup[0]);

        // System.out.println(data == backup);
        // System.out.println(data.equals(backup));
        // System.out.println(Arrays.equals(data, backup));
        // // System.out.println(Arrays.deepEquals(data, backup));

        // int[][] data = { { 1, 2 }, { 3, 4 } };
        // int[][] backup;
        // backup = data.clone();
        // data[0][0] = 11;
        // System.out.println(backup[0][0]);

        // System.out.println(data == backup);
        // System.out.println(data.equals(backup));
        // System.out.println(Arrays.equals(data, backup));
        // // System.out.println(Arrays.deepEquals(data, backup));

        int[][] data = { { 1, 2 }, { 3, 4 } };
        int[][] backup = new int[data.length][];
        backup[0] = data[0].clone();
        backup[1] = data[1].clone();
        data[0][0] = 11;
        System.out.println(backup[0][0]);

        System.out.println(data == backup);
        System.out.println(data.equals(backup));
        System.out.println(Arrays.equals(data, backup));
        // System.out.println(Arrays.deepEquals(data, backup));
    }
}
